File: 2022/2022W17/PreppindataW17.py
import pandas as pd
import re

xlsx = pd.ExcelFile(r"C:\Data\PreppinData\2022W17 Input.xlsx")

stream = pd.read_excel(xlsx, 'Streaming')
stream['timestamp'] = stream['t'].apply(lambda x: pd.to_datetime(re.sub('[A-Z]',' ',x)))
stream['location'] = stream['location'].str.replace('Edin.*','Edinburgh',regex=True)
stream['content_type'] = stream.apply(lambda x: 'Primary' if re.search('(London|Cardiff|Edinburgh)', x['location']) else 
                                      'Secondary' if pd.isna(x['content_type']) else x['content_type'], axis=1)
stream = stream.groupby(['userID', 'timestamp', 'location', 'content_type'], as_index=False).agg({'duration':'sum'})

stream['dummy_location'] = stream.apply(lambda x: 'dummy' if x['content_type']=='Primary' else x['location'], axis=1)

stream['min timestamp'] = stream['timestamp'].groupby([stream['userID'], stream['content_type'], stream['dummy_location']]).transform('min')
stream['Month'] = stream['min timestamp'].apply(lambda x: x.strftime('%m %Y'))

final = pd.merge(stream, pd.read_excel(xlsx, 'Avg Pricing').rename(columns={'Content_Type':'content_type'}), on=['Month', 'content_type'], how='left')
final['Avg_Price'] = final.apply(lambda x: 14.98 if x['content_type']=='Preserved' else x['Avg_Price'], axis=1) 

final = final[['userID', 'timestamp', 'location', 'content_type', 'duration', 'Avg_Price']]

File: 2022/2022W17/preppindataW17.R
library(readxl)
library(dplyr)
library(stringr)

xlsx <- "C:/Data/PreppinData/2022W17 Input.xlsx"

final <-  read_excel(xlsx, 'Streaming') %>%
  mutate('location' = str_replace(`location`, '^Edin.*', 'Edinburgh'), 
         'timestamp' = as.POSIXct(str_replace_all(`t`, '[A-Z]', ' ')),
         'content_type' = if_else(str_detect(`location`,'(London|Cardiff|Edinburgh)'),'Primary',
                                  if_else(is.na(`content_type`),'Secondary', `content_type`))) %>%
  group_by(`userID`, `timestamp`, `location`, `content_type`) %>%
  summarise('duration' = sum(`duration`), .groups='drop') %>%
  mutate('dummy_location' = if_else(`content_type`=='Primary', 'dummy',`location`)) %>%
  group_by(`userID`, `dummy_location`, `content_type`) %>%
  mutate('Month' = format(min(`timestamp`), '%m %Y')) %>%
  merge(read_excel(xlsx, 'Avg Pricing') %>% rename('content_type'='Content_Type'), by=c('Month','content_type'), all.x = TRUE) %>%
  mutate('Avg_Price' = if_else(`content_type`=='Preserved', 14.98, `Avg_Price`)) %>%
  select(c('userID', 'timestamp', 'location', 'content_type', 'duration', 'Avg_Price'))
  
View(final)
